Tests/PackageLoading/InMemoryConventionTests.swift

```diff
@@ -0,0 +1,174 @@
+    ///     - contents: Contents of the file. Empty by default.
+    ///
+    /// - Throws: FileSystemError
+    mutating func create(_ file: AbsolutePath, contents: ByteString = ByteString()) throws {
+        // Auto create the tree.
+        try createDirectory(file.parentDirectory, recursive: true)
+        try writeFileContents(file, bytes: contents)
+    }
+
+    /// Create multiple empty files on the filesystem while recursively creating the parent directory tree.
+    ///
+    /// - Parameters:
+    ///     - files: Paths of empty files to create.
+    ///
+    /// - Throws: FileSystemError
+    mutating func createEmptyFiles(_ files: AbsolutePath...) throws {
+        // Auto create the tree.
+        for file in files {
+            try createDirectory(file.parentDirectory, recursive: true)
+            try writeFileContents(file, bytes: ByteString())
+        }
+    }
+}
+
+/// Loads a package using PackageBuilder at the given path. If package description is provided it is used,
+/// otherwise a default package with name of directory is passed to manifest.
+///
+/// - Parameters:
+///     - path: Directory where the package is located.
+///     - in: FileSystem in which the package should be loaded from.
+///     - package: PackageDescription instance to use for loading this package.
+///
+/// - Throws: ModuleError, ProductError
+private func loadPackage(_ path: AbsolutePath, in fs: FileSystem, package: PackageDescription.Package? = nil) throws -> PackageModel.Package {
+    let pkg: PackageDescription.Package
+    if let package = package {
+        pkg = package
+    } else {
+        pkg = Package(name: path.basename)
+    }
+    let manifest = Manifest(path: path.appending(component: Manifest.filename), url: "", package: pkg, products: [], version: nil)
+    let builder = PackageBuilder(manifest: manifest, path: path, fileSystem: fs)
+    return try builder.construct(includingTestModules: true)
+}
+
+extension PackageBuilderTester {
+    @discardableResult
+    init(_ path: AbsolutePath, in fs: FileSystem, package: PackageDescription.Package? = nil, result: (PackageBuilderTester) -> Void) {
+        do {
+            self = .package(try loadPackage(path, in: fs, package: package))
+        } catch {
+            self = .error(error)
+        }
+        result(self)
+    }
+}
+
+extension PackageModel.Package {
+    var allModules: [Module] {
+        return modules + testModules
+    }
+}
+
+enum PackageBuilderTester {
+    case package(PackageModel.Package)
+    case error(Swift.Error)
+
+    func checkError(_ str: String) {
+        switch self {
+        case .package(let package):
+            XCTFail("\(package) did not have any error")
+        case .error(let error):
+            XCTAssertEqual(String(error), str)
+        }
+    }
+
+    func checkModule(_ name: String, result: ((ModuleResult) -> Void)? = nil) {
+        guard case .package(let package) = self else {
+            return XCTFail("Error: \(self)")
```

Comment:
Can we make these failure messages more verbose, something like `expected module \(name) but package did not load`?